xtimer: Allow arbitrary timer frequency, second attempt

gebart opened the pull request:
Significantly simplified version of #3990, using the ideas from #4002.
The public API of xtimer now uses two new types `xtimer_ticks32_t`, and `xtimer_ticks64_t`, to catch any missing usec/sec/nanosec to tick conversions. Internally, all timer scheduling and offsets are using hardware ticks instead of microseconds.
All internal xtimer functions use untyped (i.e. `uint32_t`, `uint64_t`) ticks and are prefixed with underscore to avoid mistakes. This extra level of abstraction is only a thin wrapper and the conversions between ticks and microseconds are inlined and will become a no-op if the hardware timer also runs at 1 MHz.

`xtimer_usleep_until` has been renamed `xtimer_periodic_wakeup`, to better reflect the purpose (and since it is now using ticks and not microseconds as parameters). This change was forked off into #5612.

~~Based on #5294, #5607.~~

Marked as WIP until all tests and drivers have been adapted to use the new `xtimer_ticks*_t` type.

gebart commented:
Adding the CI build flag to assist in finding missing API updates..

gebart commented:
Verified that the inline function for usec->tick conversion of a compile time constant generates no extra code via manual inspection of assembly output. Optimization level `-O0` generates function calls for the compile time constants. `-O1`, `-Os`, and `-Og` all precompute the constant to store it as ticks directly in the binary instead of storing as microseconds.

Conversion of run time variable values will of course generate conversion instructions since it is not possible to rescale them during compile time.

gebart commented:
#5612 was split off from this PR as a separate change.

gebart commented:
postponing to release 2016.10

immesys commented:
We can do some testing of this if you want, we need this for our low power patches on SAMR21

gebart commented:
@immesys yes please! I appreciate any support.
I will rebase and fix the conflicts.

gebart commented:
rebased, doing a CI build to see whether the rebase went well.

immesys commented:
Great, thanks! We'll work on this today. We have a SAMR21 based mote running RIOT at ~10uA average current, but xtimer does not like the duty cycling at the moment, so we are trying to get it to run off the low power 32khz clock that runs during sleep.

immesys commented:
Feedback so far: everything appears to work. We're pretty excited, this was one of the last missing pieces we needed for low power. It's a bit of a nuisance converting ticks to us wherever xtimer_now is called, but if I understand it correctly, the idea is that that is temporary? Eventually the drivers and various other callers of xtimer_now would change to use ticks natively instead of us?

immesys commented:
Do you think this will be merged for the next release? If so I need to change some other PRs I have open so they will work :p

immesys commented:
Also, as an aside, I had some trouble inside xtimer_periodic_wakeup with ticks vs uint32 [here](https://github.com/gebart/RIOT/blob/pr/xtimer-ticks/sys/xtimer/xtimer.c#L106)

immesys commented on sys/xtimer/xtimer.c:
```diff
@@ -58,15 +58,15 @@ void _xtimer_sleep(uint32_t offset, uint32_t long_offset)
 
     uint32_t target = (*last_wakeup) + period;
-    uint32_t now = xtimer_now();
+    uint32_t now = _xtimer_now();
```
You need the underscore variant on L106 xtimer_spin() and L115 xtimer_now() as well

immesys commented:
Ok I've hit a couple problems. I'll pick this up tomorrow, but I am getting stuck inside spin inside _xtimer_set64.

immesys commented:
@gebart, the problems I had were unrelated and happened during rebasing. My review is that everything works once you fix the L115 and L106 as I mentioned above. I'd like to help get this merged, so let me know how I can help.

gebart commented:
@immesys Thank you for testing this. I'll update the PR with the above mentioned fixes.

kYc0o commented:
There lots of errors in many boards, mostly due to the API change... is it possible to fix them?

gebart commented:
Yeah, I need to go through all the users of xtimer now and update. Feel
free to post a PR against this PR branch of you want to assist.

gebart commented:
Fixed all errors and warnings, just needs a squash and an ACK

kYc0o commented:
It seems very good for me, thus I'd give my ACK. Maybe @kaspar030 wants to give some feedback?

kaspar030 commented:
Looks good to me! Nice PR. 

Please squash!

(I wish simple typedefs could be marked as "strict" so there was no need for the structs and accompanying functions...)

I didn't do any testing, though. Should we make a checkbox list for the most important ones? ;)

gebart commented:
should I squash all of the changes in drivers, sys, cpu, boards, tests into a single commit?

immesys commented:
Personally I would advise against it. If this PR changes some behavior for other people, and they are tracking it down it is good to be able to bisect or cherry pick through different commits.

kYc0o commented:
@immesys would you like to propose which commits shouldn't be squashed? Because several are just some fixes to get Murdock happy or small typos, etc.

I'd like to merge this before the release BTW...

immesys commented:
Yeah, I don't want to be the one to hold things up. I'd be happy with a squash.

immesys commented:
Incidentally @gebart, I added some code that uses xtimer that was merged after your rebase, and I don't want to hold you up on another rebase so I opened another PR for it.

kYc0o commented:
@gebart Can you rebase and squash? I think this can be in for this release. What do you think @miri64 ?

miri64 commented:
> What do you think @miri64 ?

If it goes in before feature freeze and higher priority reviews aren't blocked by this effort, who am I to judge. ;-)

immesys commented:
Is there anything I can do to help get this in the release?

gebart commented:
Sorry about the slow replies, things are quite busy over here at the moment. However, I have rebased and squashed this PR and now we need to see if there have been any changes since last rebase which breaks it. Waiting for Murdock.

miri64 commented:
There are some `xtimer_now()` -> `xtimer_now_usec()` missing.

miri64 commented:
It's just the new `driver_lpd8808` introduces in #5732 as far as I can see.

miri64 commented:
Now that `gnrc_sock` is merged that and its tests need some adaptation to, I guess.

gebart commented:
I'll update that as soon as possible, likely later tonight. Thanks for the
ping

On Oct 27, 2016 11:57, "Martine Lenders" <notifications@github.com> wrote:

> It's just the new driver_lpd8808 introduces in #5732
> <https://github.com/RIOT-OS/RIOT/pull/5732> as far as I can see.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/5608#issuecomment-256599149>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AATYQiizfE2Z8v3OTgUwEpDw_FhCwN8hks5q4HWagaJpZM4JFahM>
> .
>

miri64 commented:
Postponed due to feature freeze

immesys commented:
Will this be merged now that the release is out?

immesys commented:
@gebart this no longer merges into master

gebart commented:
rebased

kYc0o commented:
Can you squash your fixup? I think everything is OK now so we could have this PR merged soon :)

gebart commented:
@kYc0o 
squashed the fixup

kYc0o commented:
And go!

OlegHahm commented:
Wow, cool!

immesys commented:
Awesome! Thank you all for your work here
